ARTEMIS-3657 refactor address docs (#3930)
apache/activemq-artemis: docs/user-manual/en/address-model.md

# Address Model

Every messaging protocol and API that Apache ActiveMQ Artemis supports defines
a different set of messaging resources.

 - JMS uses _queues_ and _topics_
 - STOMP uses generic _destinations_
 - MQTT uses _topics_
 - AMQP uses generic _nodes_

In order to deal the the unique semantics and use-cases for each of these the
broker has a flexible and powerful address model based on the following _core_
set of resources:

 - **address**
 - **queue**
 - **routing type**

### Address

Messages are *sent* to an address. An address is given a unique name, a routing

> type, and zero or more queues.

Author: I have routing type in here to explain a configuration like this:
```xml
<address name="myTopic">
    <multicast/>
</address>
```
Here the address has a routing type and 0 queues.

### Queue

Messages are *consumed* from a queue. A queue is bound to an address. There can
be zero or more queues bound to one address. When a message is sent to an
address it is routed to one or more of its queues based on the configured

> routing type.

Author: Done.

The name of the queue must be _globally_ unique. For example, you can't have a
queue named `q1` on address `a1` and also a queue named `q1` address `a2`.

### Routing Type

A routing type determines how messages are routed from an address to the
queue(s) bound to that address. Two different routing types are supported,
**anycast** and **multicast**.

If you want your messages routed to...|Use this routing type...
---|---

> a single queue on the address.|anycast
> every queue on the address.|multicast

Reviewer: ```suggestion
a single anycast queue on the address.|anycast
every multiple queue on the address.|multicast
```
When the broker routes a message takes into account 2 routing types: the one of the message and the one configured in the queue. I don't know if this is important for users and how to describe it in a clear way. Indeed the suggestion is just a draft to clarify this idea.

> **Note:**
> 

> > It is possible to define more than one routing type per address, but this

Reviewer: ```suggestion
> It is possible to define queues with a different routing type for the same address, but this
```

Author: Done.

> > typically results in an anti-pattern and is therefore not recommended.

Author: I honestly can't think of a use-case for this. Can you? I'd be happy to add an explanation of straight-forward use-case if one exists.

> > typically results in an anti-pattern and is therefore not recommended.

Reviewer: I have seen more than one routing type per address only as workaround for the following case:
A JMS producer sends messages to `TEST.QUEUE` address while 2 consumers receive messages from 2 FQQN queues `fqqn://TEST.QUEUE::QUEUE1` and `fqqn://TEST.QUEUE::QUEUE2`.
```
<address name="TEST.QUEUE">
    <anycast>
       <queue name="QUEUE1" />
       <queue name="QUEUE2" />
    </anycast>
    <multicast>
       <queue name="TEST.QUEUE" />
    </multicast>
</address>
```

If the multicast queue `TEST.QUEUE` isn't defined the broker will auto-create it as anycast queue distributing messages to `TEST.QUEUE` too.

> > typically results in an anti-pattern and is therefore not recommended.

Reviewer: Sounds good

a producer sends a message to `address.foo` it is then routed to `q1` and
finally dispatched to one of the consumers.

![Anycast](images/anycast.png)
Figure 1. Anycast 

This is what the configuration for this use-case would look like in 
`etc/broker.xml`:

```xml
<addresses>
   <address name="address.foo">
      <anycast>
         <queue name="q1"/>
      </anycast>
   </address>
</addresses>
```

For most protocols and APIs which support this kind of use-case (e.g. JMS, 
AMQP, etc.) it is customary to use the _same name_ when sending and consuming
messages. In that case you'd use a configuration like this:

```xml
<addresses>
   <address name="orderQueue">
      <anycast>
         <queue name="orderQueue"/>
      </anycast>
   </address>
</addresses>
```

### Multicast

The most common use-case for multicast semantics, sometimes referred to as 
[publish/subscribe](messaging-concepts.md#publish-subscribe) or "pub/sub",
involves each application receiving every message sent to an address. Multiple
applications consuming from a JMS topic is a classic example of this use-case.
MQTT subscriptions is another supported example of multicast semantics.

> In this use-case the broker is configured with an address, `anycast.foo` using

Author: Done.

```xml
<addresses>
   <address name="address.foo">
      <multicast/>
   </address>
</addresses>
```
Define `<multicast/>` with no queues and the broker will automatically create
queues for each subscription when the consumers connect to `address.foo`. Then
when a message is sent to `address.foo` it will be routed to each queue for
each subscriber and therefore each subscriber will get every message. These
queues are often referred to as _subscription queues_ for obvious reasons.

These subscription queues are typically named based on the semantics of the
protocol used to create them. For example, JMS supports durable and non-durable
subscriptions. The queue for a non-durable subscription is named with a UUID,
but the queue used for a durable subscription is named according to the JMS
"client ID" and "subscription name." Similar conventions are used for AMQP,
MQTT, STOMP, etc.

## Advanced Manual Configuration

### Fully Qualified Queue Names

In most cases it’s not necessary or desirable to statically configure the
aforementioned subscription queues. However, there are scenarios where a user
may want to statically configure a subscription queue and later connect to that

> queue directly using a **Fully Qualified Queue Name** (i.e. FQQN).

Author: Done.

An FQQN uses a special syntax to specify *both* the address and the queue so
that applications using protocols and APIs which don't natively understand the
address/queue separation (e.g. AMQP, JMS, etc.) can send messages or subscribe
*directly* to a queue rather than being limited to the address. Applications
simply need to use the address name and the queue name separated by `::` (e.g.
`address::queue`).

In this example, the address `a1` is configured with two queues: `q1`, `q2` as
shown in the configuration below.

```xml
<addresses>
   <address name="a1">
      <multicast>
         <queue name="q1" />
         <queue name="q2" />
      </multicast>
   </address>
</addresses>
```

Here's a snippet of Java code using JMS which demonstrates the FQQN syntax:

```java
String FQQN = "a1::q1";

> Queue q1 session.createQueue(FQQN);

Author: Done.
